x/tally/keeper/payout.go

```diff
@@ -35,27 +58,114 @@ func (k Keeper) CalculateCommitterPayouts(ctx sdk.Context, req types.Request, ga
+
+		adjPayoutAmt := burnRatio.MulInt(payoutAmt).TruncateInt()
+		burnAmt = payoutAmt.Sub(adjPayoutAmt)
+		payoutAmt = adjPayoutAmt
+	}
+
+	totalBurn := math.ZeroInt()
+	dists := make([]types.Distribution, len(executors))
+	for i, executor := range executors {
+		if burnAmt.IsPositive() {
+			totalBurn = totalBurn.Add(burnAmt)
 		}
+		dists[i] = types.NewExecutorReward(executor, payoutAmt)
+
+		ctx.EventManager().EmitEvent(sdk.NewEvent(types.EventTypeExecutorRewardUniform,
+			sdk.NewAttribute(types.AttributeExecutor, executor),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, payoutAmt.String()),
+		))
 	}
-	result.Messages = distMsgs
-	return result, nil
+	return dists, adjGasUsed * uint64(replicationFactor), totalBurn
 }
 
-// TODO: This will become more complex when we introduce incentives.
-func calculateExecGasUsed(reveals []types.RevealBody) uint64 {
-	var execGasUsed uint64
-	for _, reveal := range reveals {
-		execGasUsed += reveal.GasUsed
+// CalculateDivergentPayouts calculates payouts for the executors given their
+// divergent gas reports. It also returns the total execution gas consumption.
+// It assumes that the i-th executor is the one who revealed the i-th reveal.
+func (k Keeper) CalculateDivergentPayouts(ctx sdk.Context, executors []string, gasReports []uint64, execGasLimit uint64, replicationFactor uint16, gasPrice math.Int, reducedPayout bool) ([]types.Distribution, uint64, math.Int) {
+	adjGasUsed := make([]uint64, len(gasReports))
+	var lowestGasUsed uint64
+	var lowestReporterIndex int
+	for i, gasReport := range gasReports {
+		adjGasUsed[i] = min(gasReport, execGasLimit/uint64(replicationFactor))
+		if i == 0 || adjGasUsed[i] < lowestGasUsed {
+			lowestReporterIndex = i
+			lowestGasUsed = adjGasUsed[i]
+		}
+	}
+	medianGasUsed := median(adjGasUsed)
+	totalGasUsed := medianGasUsed*uint64(replicationFactor-1) + min(lowestGasUsed*2, medianGasUsed)
+	totalShares := medianGasUsed*uint64(replicationFactor-1) + lowestGasUsed*2
+	lowestPayout := gasPrice.Mul(math.NewIntFromUint64(lowestGasUsed * 2 * totalGasUsed / totalShares))
+	normalPayout := gasPrice.Mul(math.NewIntFromUint64(medianGasUsed * totalGasUsed / totalShares))
+
+	burnRatio := math.LegacyNewDecWithPrec(2, 1) // burn 20% in case of reduced payout
+
+	totalBurn := math.ZeroInt()
+	dists := make([]types.Distribution, len(executors))
+	for i, executor := range executors {
+		payout := normalPayout
+		if i == lowestReporterIndex {
+			payout = lowestPayout
+		}
+		if reducedPayout {
+			adjPayout := burnRatio.MulInt(payout).TruncateInt()
+			totalBurn = totalBurn.Add(payout.Sub(adjPayout))
+			payout = adjPayout
+		}
+		dists[i] = types.NewExecutorReward(executor, payout)
+
+		ctx.EventManager().EmitEvent(sdk.NewEvent(types.EventTypeExecutorRewardUniform,
+			sdk.NewAttribute(types.AttributeExecutor, executor),
+			sdk.NewAttribute(sdk.AttributeKeyAmount, payout.String()),
+		))
+	}
+	return dists, totalGasUsed, totalBurn
```

Comment:
Note that in the current implementation, the total gas used does not change even in case of reduced payout.